File: src/main/java/no/fint/provider/springer/model/UtdanningKodeverkRepository.java
package no.fint.provider.springer.model;

import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.Event;
import no.fint.event.model.ResponseStatus;
import no.fint.event.model.Status;
import no.fint.model.resource.FintLinks;
import no.fint.model.resource.utdanning.kodeverk.ElevkategoriResource;
import no.fint.model.resource.utdanning.kodeverk.FravarstypeResource;
import no.fint.model.resource.utdanning.kodeverk.KarakterskalaResource;
import no.fint.model.resource.utdanning.kodeverk.SkoleeiertypeResource;
import no.fint.model.utdanning.kodeverk.KodeverkActions;
import no.fint.provider.springer.storage.SpringerRepository;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Repository;

import javax.annotation.PostConstruct;
import java.util.EnumMap;
import java.util.Set;
import java.util.stream.Collectors;

import static no.fint.model.utdanning.kodeverk.KodeverkActions.*;

@Slf4j
@Repository
public class UtdanningKodeverkRepository extends SpringerRepository {

    private final EnumMap<KodeverkActions, Class<? extends FintLinks>> actions = new EnumMap<>(KodeverkActions.class);

    @PostConstruct
    public void init() {
        actions.put(GET_ALL_ELEVKATEGORI, ElevkategoriResource.class);
        actions.put(GET_ALL_FRAVARSTYPE, FravarstypeResource.class);
        actions.put(GET_ALL_KARAKTERSKALA, KarakterskalaResource.class);
        actions.put(GET_ALL_SKOLEEIERTYPE, SkoleeiertypeResource.class);
    }

    @Override
    public void accept(Event<FintLinks> response) {
        if (!StringUtils.contains(response.getSource(), "utdanning")) {
            log.info("Skipping {} from {}", response.getAction(), response.getSource());
            return;
        }
        KodeverkActions action = valueOf(response.getAction());
        if (actions.containsKey(action)) {
            query(actions.get(action), response);
        } else {
            response.setStatus(Status.ADAPTER_REJECTED);
            response.setResponseStatus(ResponseStatus.REJECTED);
            response.setStatusCode("INVALID_ACTION");
            response.setMessage("Invalid action: " + action);
        }
    }

    @Override
    public Set<String> actions() {
        return actions.keySet()
                .stream()
                .map(Enum::name)
                .collect(Collectors.toSet());
    }
}

File: src/main/kotlin/no/fint/provider/springer/model/UtdanningKodeverkRepository.kt
package no.fint.provider.springer.model

import jakarta.annotation.PostConstruct
import no.fint.event.model.Event
import no.fint.event.model.ResponseStatus
import no.fint.event.model.Status
import no.fint.provider.springer.storage.SpringerRepository
import no.fint.provider.springer.storage.Wrapper
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.resource.utdanning.kodeverk.*
import no.novari.fint.model.utdanning.kodeverk.KodeverkActions
import org.apache.commons.lang3.StringUtils
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Repository
import java.util.*
import java.util.stream.Collectors

@Repository
class UtdanningKodeverkRepository(
    mongoTemplate: MongoTemplate,
    wrapper: Wrapper
) : SpringerRepository(mongoTemplate, wrapper) {

    private val actions: EnumMap<KodeverkActions, Class<out FintLinks>> = EnumMap(KodeverkActions::class.java)

    @PostConstruct
    fun init() {
        actions[KodeverkActions.GET_ALL_ELEVKATEGORI] = ElevkategoriResource::class.java
        actions[KodeverkActions.GET_ALL_FRAVARSTYPE] = FravarstypeResource::class.java
        actions[KodeverkActions.GET_ALL_KARAKTERSKALA] = KarakterskalaResource::class.java
        actions[KodeverkActions.GET_ALL_SKOLEEIERTYPE] = SkoleeiertypeResource::class.java
        actions[KodeverkActions.GET_ALL_SKOLEAR] = SkolearResource::class.java
        actions[KodeverkActions.GET_ALL_TERMIN] = TerminResource::class.java
    }

    override fun accept(response: Event<FintLinks>) {
        if (!StringUtils.contains(response.source, "utdanning")) {
            return
        }
        val action = KodeverkActions.valueOf(response.action)
        val type = actions[action]
        if (type != null) {
            query(type, response)
        } else {
            response.status = Status.ADAPTER_REJECTED
            response.responseStatus = ResponseStatus.REJECTED
            response.statusCode = "INVALID_ACTION"
            response.message = "Invalid action: $action"
        }
    }

    override fun actions(): Set<String> =
        actions.keys.stream()
            .map { it.name }
            .collect(Collectors.toSet())
}
